url80587fc7d5_dennervidal_CloudSim_DVFS_tgz-pJ8-Examples_Dvfs_example_simpleJ8/src/org/cloudbus/cloudsim/network/GraphReaderBrite.java

```diff
@@ -0,0 +1,204 @@
+/*
+ * Title:        CloudSim Toolkit
+ * Description:  CloudSim (Cloud Simulation) Toolkit for Modeling and Simulation of Clouds
+ * Licence:      GPL - http://www.gnu.org/copyleft/gpl.html
+ *
+ * Copyright (c) 2009-2012, The University of Melbourne, Australia
+ */
+
+package org.cloudbus.cloudsim.network;
+
+import java.io.BufferedReader;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.StringTokenizer;
+
+/**
+ * This class is just an file-reader for the special brite-format! the brite-file is structured as
+ * followed: Node-section: NodeID, xpos, ypos, indegree, outdegree, ASid, type(router/AS)
+ * Edge-section: EdgeID, fromNode, toNode, euclideanLength, linkDelay, linkBandwith, AS_from, AS_to,
+ * type
+ * 
+ * @author Thomas Hohnstein
+ * @since CloudSim Toolkit 1.0
+ */
+public class GraphReaderBrite implements GraphReaderIF {
+
+	private static final int PARSE_NOTHING = 0;
+
+	private static final int PARSE_NODES = 1;
+
+	private static final int PARSE_EDGES = 2;
+
+	private int state = PARSE_NOTHING;
+
+	private TopologicalGraph graph = null;
+
+	/**
+	 * this method just reads the file and creates an TopologicalGraph object
+	 * 
+	 * @param filename name of the file to read
+	 * @return created TopologicalGraph
+	 * @throws IOException
+	 */
+	@Override
+	public TopologicalGraph readGraphFile(String filename) throws IOException {
+
+		graph = new TopologicalGraph();
+
+		// lets read the file
+		FileReader fr = new FileReader(filename);
+		BufferedReader br = new BufferedReader(fr);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *br*. The resource is referenced at lines 51, 57. The resource is closed at line 87. There are other execution paths that do not contain closure statements, for example, when BufferedReader.readLine throws an exception. Ensure *br* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).